src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs

```diff
@@ -0,0 +1,129 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace Microsoft.Identity.Client.ApiConfig
+{
+    // TODO: we should only have values in here publicly that are useful for ALL acquire token usages
+    // (e.g. WithExtraQueryParameters).  Which ones don't make the cut and need to be moved up the 
+    // inheritance hierarchy?
+    /// <summary>
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TParams"></typeparam>
+    public abstract class AbstractAcquireTokenParameterBuilder<T, TParams>
+        where T : AbstractAcquireTokenParameterBuilder<T, TParams>
+        where TParams : IAcquireTokenCommonParameters
+    {
+        internal AcquireTokenParameters Parameters { get; } = new AcquireTokenParameters();
+
+        /// <summary>
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <returns></returns>
+        protected T WithScopes(IEnumerable<string> scopes)
+        {
+            Parameters.Scopes = scopes;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="loginHint"></param>
+        /// <returns></returns>
+        public T WithLoginHint(string loginHint)
+        {
+            Parameters.LoginHint = loginHint;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="account"></param>
+        /// <returns></returns>
+        public T WithAccount(IAccount account)
+        {
+            Parameters.Account = account;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="extraQueryParameters"></param>
+        /// <returns></returns>
+        public T WithExtraQueryParameters(IDictionary<string, string> extraQueryParameters)
+        {
+            Parameters.ExtraQueryParameters = new ReadOnlyDictionary<string, string>(extraQueryParameters ?? new Dictionary<string, string>());
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="extraScopesToConsent"></param>
+        /// <returns></returns>
+        public T WithExtraScopesToConsent(IEnumerable<string> extraScopesToConsent)
+        {
+            Parameters.ExtraScopesToConsent = extraScopesToConsent;
+            return (T)this;
+        }
+
+        /// <summary>
+        ///     TODO: replicate the options here that we have in ApplicationBuilder for an AuthorityInfo class?
+        /// </summary>
+        /// <param name="authorityUri"></param>
+        /// <returns></returns>
+        public T WithAuthorityOverride(string authorityUri)
```

Comment:
the question here is if we'd want to have more overrides like WithAzureADAuthorityOverride with clouds / audience?

Reply:
Yes, I think we need to look at this the same way we do authorities in the app builder scenario (thus the TODO in the summary of this method).  I can look at taking what we have finalized in the app builder scenario and putting it here as well.

---
In reply to: [246267784](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246267784) [](ancestors = 246267784)